File: math/permutations/permutations.cpp
/**
 * by Jaroslaw Hirniak
 * 07 November 2014
 */

#define DEBUG

#include <iostream>
#include <string>
#include <set>

using namespace std;

/**
 * creates permutation by iterating over existing permutations
 * and inserting char c in each possible position
 */
set<string> permute(char c, set<string> perms) {
  set<string> newPerms {};
  
  for (auto it: perms) {
    for (size_t i = 0; i <= it.size(); ++i) {
      newPerms.insert(it.substr(0, i) + c + it.substr(i));
    }
  }
  
  return newPerms;
}

/**
 * initializes recursive permutation function
 */
set<string> getPermutations(string str) {
    if(!str[0]) {
          return set<string> {""};
            }
      return permute(str[0], getPermutations(str.substr(1)));
}

int main() {
  #ifdef DEBUG
    if (freopen("input.txt", "r", stdin) == 0) {
      cout << "Loading test file failed" << endl;
    } else {
      string str;
      while (cin >> str) {
        set<string> permutations = getPermutations(str);
        cout << "***\n";
        for (auto it: permutations) {
          cout << "\t" << it << endl;
        }
      }
      cout << "***\n";
    }
  #endif

  return 0;
}

File: math/permutations/permutations.py


def permutations(items):
    if len(items) <= 1:
        yield items
    else:
        for permutation in permutations(items[1:]):
            for i in range(len(items)):
                mutator = items[0] if type(items[0]) == str else [items[0]]
                yield permutation[:i] + mutator + permutation[i:]


tests = [
    'abcd',
    [1, 2, 3]
]


if __name__ == '__main__':
    for test in tests:
        perms = list(permutations(test))
        print('Permutations ({}):\n{}'.format(len(perms), perms))
